File: solar-system-3d/Planet.js
class Planet {
  constructor(radius, distance, color, orbitSpeed = random(0.01, 0.1)) {
    this.radius = radius;
    this.distance = distance;
    this.vector = p5.Vector.random3D().mult(this.distance);
    this.color = color;
    this.orbitSpeed = orbitSpeed;
    this.angle = random(0, 3.14);
    this.moons = [];
  }
  
  show() {
    // temporarily reset drawing state
    push();

    const crossV = this.vector.cross(createVector(1, 0, 1));
    
    // length of the vector must be > 0
    if (crossV.x || crossV.y || crossV.z) {
      rotate(this.angle, crossV);
    }
    
    // arrange distance before drawing
    translate(this.vector.x, this.vector.y, this.vector.z);
    
    fill(this.color.r, this.color.g, this.color.b);
    noStroke();
    sphere(this.radius);
    
    if (this.moons.length > 0) {
      for (const moon of this.moons) {
        moon.show();
      }
    }
    
    pop();
  }
  
  generateMoons(count, level) {    
    for (let i = 0; i < count; i++) {
      const newRadius = this.radius / (level * 3);
      const newDistance = random((newRadius + this.radius) * 2, (newRadius + this.radius) * 3);
      const newSpeed = random(-0.1, 0.1);
      const newColor = {r: random(50, 200), g: random(100, 255), b: random(100, 255)};
      
      this.moons[i] = new Planet(newRadius, newDistance, newColor, newSpeed);
      
      if (level < 2) {
        this.moons[i].generateMoons(random(0, 4), level + 1);
      }
    }
  }
  
  orbit() {
    this.angle += this.orbitSpeed;
    
    if (this.moons.length > 0) {
      for (const moon of this.moons) {
        moon.orbit();
      }
    }
  }
}

File: sketches/solar_system_3d/Planet.py
class Planet(object):
    def __init__(self, radius, distance, color, speed=random(0.01, 0.1)):
        self.radius = radius
        self.distance = distance
        self.vector = PVector.random3D().mult(self.distance)
        self.color = color
        self.speed = speed
        self.angle = random(TWO_PI)
        self.moons = []
        
        
    def show(self):
        push()
        v1 = PVector(self.vector.x, self.vector.y, self.vector.z)
        v2 = v1.cross(PVector(1, 0, 1))
        rotate(self.angle, v2.x, v2.y, v2.z)

        # Adjust distance before drawing
        translate(self.vector.x, self.vector.y, self.vector.z)
        noStroke()
        fill(self.color["r"], self.color["g"], self.color["b"])
        sphere(self.radius)
        
        if self.moons:
            for moon in self.moons:
                moon.show()
        
        pop()
        
        
    def generate_moons(self, count, level):
        for i in range(count):
            new_radius = self.radius / (level * 1.75)
            new_distance = random((new_radius + self.radius) * 1.25, (new_radius + self.radius) * 3)
            new_color = {"r": int(random(50, 200)), "g": int(random(100, 255)), "b": int(random(100, 255))}
            new_speed = random(-0.1, 0.1)
            
            self.moons.append(Planet(new_radius, new_distance, new_color, new_speed))

            if level < 3:
                self.moons[i].generate_moons(int(random(0, 4)), level + 1)
    
    
    def orbit(self):
        self.angle += self.speed
    
        if self.moons:
            for moon in self.moons:
                moon.orbit()
                
